precovery/brute_force.py

```diff
@@ -0,0 +1,387 @@
+import os
+
+from sqlalchemy import all_
+
+from precovery.precovery_db import PrecoveryDatabase
+
+os.environ["OMP_NUM_THREADS"] = "1"
+os.environ["OPENBLAS_NUM_THREADS"] = "1"
+os.environ["MKL_NUM_THREADS"] = "1"
+os.environ["VECLIB_MAXIMUM_THREADS"] = "1"
+os.environ["NUMEXPR_NUM_THREADS"] = "1"
+
+import logging
+import multiprocessing as mp
+import time
+from functools import partial
+from typing import Iterable, List, Optional, Tuple
+
+import numpy as np
+import pandas as pd
+from astropy.time import Time
+from sklearn.neighbors import BallTree
+
+from .healpix_geom import radec_to_healpixel, radec_to_healpixel_with_neighbors
+
+# replace this usage with Orbit.compute_ephemeris
+from .orbit import Orbit
+from .residuals import calcResiduals
+from .utils import _checkParallel, _initWorker, calcChunkSize, yieldChunks
+
+logger = logging.getLogger(__name__)
+
+__all__ = [
+    "get_observations_and_ephemerides_for_orbits",
+    "attribution_worker",
+    "attributeObservations",
+]
+
+
+def get_observations_and_ephemerides_for_orbits(
+    orbits: Iterable[Orbit],
+    mjd_start: float,
+    mjd_end: float,
+    precovery_db: PrecoveryDatabase,
+    obscode: str = "W84",
+):
+    """
+    Returns:
+    - ephemerides for the orbit list, propagated to all epochs for frames in 
+    the range (mjd_start, mjd_end)
+    - observations in the indexed PrecoveryDatabase consistent with intersecting
+    and neighboring frames for orbits propagated to each of the represented epochs
+
+    ***Currently breaks on non-NSC datasets
+        TODO propagation targets should be unique frames wrt obscode, mjd
+    """
+
+    # Find all the mjd we need to propagate to
+    all_frame_mjd = precovery_db.frames.idx.unique_frame_times()
```

Comment:
You've probably already thought about this but you can solve your comment on line 54 by changing this function to return tuples (or similar) of obscode and mjds for that obscode. Then you can remove the obscode kwarg altogether.

Reply:
Yep - in my local branch now, will push a fix.

This does break down a little with per-obs MJDs though.

Reply:
Ah yes, excellent point. We might be able to get around that later on by loading in the frames and then adding a check for if `frame.mjd != obs.mjd` then propagate to the correct epoch. Not sure how we would generalize it so we can have nelson-force and brute-force use the same logic and code.